File: queue/ts/queue.ts
interface ListNode<T> {
    value: T;
    next: ListNode<T> | null;
}

export class Queue<T> {
    private head: ListNode<T> | null = null;
    private tail: ListNode<T> | null = null;
    private length: number = 0;

    enqueue(value: T): void {
        const newNode: ListNode<T> = { value, next: null };
        if (this.isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail!.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }

    dequeue(): T | null {
        if (this.isEmpty()) {
            return null;
        }
        const value = this.head!.value;
        this.head = this.head!.next;
        if (this.head === null) {
            this.tail = null;
        }
        this.length--;
        return value;
    }

    peek(): T | null {
        return this.head ? this.head.value : null;
    }

    isEmpty(): boolean {
        return this.length === 0;
    }

    size(): number {
        return this.length;
    }

    printQueue(): void {
        let current = this.head;
        let output = "";
        while (current) {
            output += `${current.value} -> `;
            current = current.next;
        }
        output += "None";
        console.log(output);
    }
}

File: queue/py/queue.py
class ListNode:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self._length = 0

    def enqueue(self, value):
        new_node = ListNode(value)
        if self.is_empty():
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self._length += 1

    def dequeue(self):
        if self.is_empty():
            return None
        value = self.head.value
        self.head = self.head.next
        if self.head is None:
            self.tail = None
        self._length -= 1
        return value

    def peek(self):
        return self.head.value if self.head else None

    def is_empty(self):
        return self._length == 0

    def size(self):
        return self._length

    def print_queue(self):
        current = self.head
        while current:
            print(current.value, end=" -> ")
            current = current.next
        print("None")

